### PR TITLE
Fixing hps remain active & meta hp configuration

### PR DESCRIPTION
**Pull request by Louquinze**
fixing the issue that metalearning tries to use every hp defined in the csv files.

Also fixing the hps remain active bug.

### CODE REVIEW

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py [hunk 1]
```diff
@@ -139,33 +155,49 @@ def fit(
...
+                )
 
             sklearn_transf_spec = [
                 (name, transformer, feature_columns)
-                for name, transformer, feature_columns in [
-                    ("categorical_transformer", self.categ_ppl, categorical_features),
-                    ("numerical_transformer", self.numer_ppl, numerical_features),
-                    ("text_transformer", self.txt_ppl, text_features),
-                ]
+                for name, transformer, feature_columns in transformer_lst
```
Comment (mfeurer) [inline autosklearn/pipeline/components/data_preprocessing/feature_type.py:200]: Is this for-loop still necessary given the changes you did above? Or could you revert the changes above?

File: autosklearn/metalearning/input/aslib_simple.py [hunk 2]
```diff
@@ -143,13 +145,17 @@ def _read_configurations(self, filename):
...
+            hp_names = self.cs.get_hyperparameter_names()
             for line in csv_reader:
                 configuration = dict()
                 algorithm_id = line["idx"]
                 for hp_name, value in line.items():
                     if not value or hp_name == "idx":
                         continue
-
+                    if hp_name not in hp_names:
+                        # skip meta learning configuration
```
Comment (mfeurer) [inline autosklearn/metalearning/input/aslib_simple.py:156]: I believe this comment is not correct, we don't skip the configuration, we only skip the hyperparameter, right?

**Review by mfeurer — COMMENTED** [review]
Good morning. Some initial feedback to help me better understand some part of the PR.

File: test/test_pipeline/components/data_preprocessing/test_data_preprocessing_feat_type.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        truth_table = [False] * 3
+        for key in cs.get_hyperparameters_dict().keys():
+            if "text" in key.split(":")[0]:
+                truth_table[0] = True
+            elif "categorical" in key.split(":")[0]:
+                truth_table[1] = True
+            elif "numerical" in key.split(":")[0]:
+                truth_table[2] = True
+
+        self.assertEqual(sum(truth_table), 3)
```
Comment (mfeurer) [inline test/test_pipeline/components/data_preprocessing/test_data_preprocessing_feat_type.py:128]: Could we also check for the number of hyperparameters in the search space?

File: test/test_pipeline/components/data_preprocessing/test_data_preprocessing_feat_type.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                "multilabel": False,
+                "multiclass": False,
+                "target_type": "classification",
+                "signed": False,
+            },
+        )
+        for key in cs.get_hyperparameters_dict().keys():
+            self.assertNotIn("text", key.split(":")[0])
+
+        DPP = FeatTypeSplit(feat_type={"A": "categorical", "B": "string"})
```
Comment (mfeurer) [inline test/test_pipeline/components/data_preprocessing/test_data_preprocessing_feat_type.py:74]: You probably wanted numerical and string here, as you also have categorical and string below.

Comment (mfeurer) [reply test/test_pipeline/components/data_preprocessing/test_data_preprocessing_feat_type.py:74]: Numerical and string also would be a good test.

File: autosklearn/pipeline/components/classification/bernoulli_nb.py [hunk 5]
```diff
@@ -64,7 +67,9 @@ def get_properties(dataset_properties=None):
         }
 
     @staticmethod
-    def get_hyperparameter_search_space(dataset_properties=None):
+    def get_hyperparameter_search_space(
+        feat_type: Optional[FEAT_TYPE_TYPE] = None, dataset_properties=None
```
[the remark this answers is not in the document]
Comment (Louquinze) [reply autosklearn/pipeline/components/classification/bernoulli_nb.py:71]: In this caste the `get_hyperparameter_search_space` function is not using the feat_type variable

File: autosklearn/util/pipeline.py [hunk 6]
```diff
@@ -44,16 +45,18 @@ def get_configuration_space(
...
     else:
         return _get_classification_configuration_space(
-            info, include, exclude, random_state
+            datamanager, include, exclude, random_state
         )
 
 
 def _get_regression_configuration_space(
-    info: Dict[str, Any],
+    datamanager: XYDataManager,
```
[the remark this answers is not in the document]
Comment (Louquinze) [reply autosklearn/util/pipeline.py:59]: Is this a more generic type which is included in python typing ? If so sure.

File: autosklearn/pipeline/components/data_preprocessing/feature_type.py [hunk 7]
```diff
@@ -46,7 +47,7 @@ def __init__(
         exclude: Optional[Dict[str, str]] = None,
         random_state: Optional[Union[int, np.random.RandomState]] = None,
         init_params: Optional[Dict[str, Any]] = None,
-        feat_type: Optional[Dict[Union[str, int], str]] = None,
+        feat_type: Optional[FEAT_TYPE_TYPE] = None,
```
[the remark this answers is not in the document]
Comment (Louquinze) [reply autosklearn/pipeline/components/data_preprocessing/feature_type.py:50]: Probability not i think i just became inconsistent after changing all these files. I think i will go through the files and check that the order is the same for all init functions

File: test/test_metalearning/pyMetaLearn/test_metalearner.py [hunk 8]
```diff
@@ -23,14 +23,26 @@ def setUp(self):
...
+        ]:
+            pipeline = (
+                autosklearn.pipeline.classification.SimpleClassificationPipeline()
+            )
+            self.cs = pipeline.get_hyperparameter_search_space(feat_type=feat_type)
+
+            self.logger = logging.getLogger()
+            meta_base = MetaBase(self.cs, data_dir, logger=self.logger)
+            self.meta_optimizer = metalearner.MetaLearningOptimizer(
+                "233", self.cs, meta_base, logger=self.logger
```
Comment (mfeurer) [inline test/test_metalearning/pyMetaLearn/test_metalearner.py:44]: Please also check for what is actually loaded. The fact we load something is great, but we need to also ensure that we load something correct.

File: autosklearn/pipeline/base.py [hunk 9]
```diff
@@ -221,7 +221,7 @@ def set_hyperparameters(
             node_name, node = n_
 
             sub_configuration_space = node.get_hyperparameter_search_space(
-                feat_type=feat_type, dataset_properties=self.dataset_properties
+                feat_type=self.feat_type, dataset_properties=self.dataset_properties
```
Comment (mfeurer) [inline autosklearn/pipeline/base.py:224]: We should try to remove `feat_type` as an argument to this function as the `feat_type` should be already given in the constructor.

File: autosklearn/pipeline/components/classification/__init__.py [hunk 10]
```diff
@@ -126,7 +133,7 @@ def get_hyperparameter_search_space(
         for estimator_name in available_estimators.keys():
             estimator_configuration_space = available_estimators[
                 estimator_name
-            ].get_hyperparameter_search_space(dataset_properties)
+            ].get_hyperparameter_search_space(dataset_properties=dataset_properties)
```
Comment (mfeurer) [inline autosklearn/pipeline/components/classification/__init__.py:136]: Should we pass the `feat_type` here just to be sure?

File: autosklearn/util/pipeline.py [hunk 11]
```diff
@@ -27,8 +28,8 @@ def get_configuration_space(
 
     Parameters
     ----------
-    info: Dict[str, Any]
-        Information about the dataset
+    datamanager: XYDataManager
```
Comment (mfeurer) [inline autosklearn/util/pipeline.py:31]: This docstring is no longer correct.

File: autosklearn/util/pipeline.py [hunk 12]
```diff
@@ -62,8 +65,8 @@ def _get_regression_configuration_space(
...
```
Comment (mfeurer) [inline autosklearn/util/pipeline.py:68]: This one neither.

File: autosklearn/util/pipeline.py [hunk 13]
```diff
@@ -109,8 +113,8 @@ def _get_classification_configuration_space(
...
```
Comment (mfeurer) [inline autosklearn/util/pipeline.py:116]: And one more.

File: scripts/02_retrieve_metadata.py [hunk 14]
```diff
@@ -178,6 +178,11 @@ def write_output(outputs, configurations, output_dir, configuration_space, metri
             fh.write("%s: %s\n" % (key, description[key]))
 
 
+class DummyDatamanager():
+    def __init__(self, info):
+        self.info = info
+        self.feat_type = {"A1": "numerical"}
```
Comment (mfeurer) [inline scripts/02_retrieve_metadata.py:184]: This appears to be a bad idea. This should be created per task to create correct configuration spaces when reading in configurations from disk/meta-learning.

File: test/fixtures/ensembles.py [hunk 15]
```diff
@@ -42,9 +42,15 @@ def _make(
         seed: Union[int, None, np.random.RandomState] = DEFAULT_SEED,
     ) -> VotingClassifier:
         assert not (X is None) ^ (y is None)
-
         if not models:
-            models = [MyDummyClassifier(config=1, random_state=seed) for _ in range(5)]
+            models = [
+                MyDummyClassifier(
+                    feat_type={i: "numerical" for i in range(X.shape[1])},
```
Comment (mfeurer) [inline test/fixtures/ensembles.py:48]: How do we know that all features are numerical?

File: test/fixtures/ensembles.py [hunk 16]
```diff
@@ -81,7 +87,14 @@ def _make(
         assert not (X is None) ^ (y is None)
 
         if not models:
-            models = [MyDummyRegressor(config=1, random_state=seed) for _ in range(5)]
+            models = [
+                MyDummyRegressor(
+                    feat_type={i: "numerical" for i in range(X.shape[1])},
```
Comment (mfeurer) [inline test/fixtures/ensembles.py:92]: Again here.

File: test/test_metalearning/pyMetaLearn/test_metalearner.py [hunk 17]
```diff
@@ -23,14 +23,27 @@ def setUp(self):
...
+            )
+            self.cs = pipeline.get_hyperparameter_search_space(feat_type=feat_type)
+            # print(self.cs.get_default_configuration())
+
+            self.logger = logging.getLogger()
+            meta_base = MetaBase(self.cs, data_dir, logger=self.logger)
+            self.meta_optimizer = metalearner.MetaLearningOptimizer(
+                "233", self.cs, meta_base, logger=self.logger
+            )
+            self.assertEqual(len(self.meta_optimizer.configuration_space), cs_size)
```
Comment (mfeurer) [inline test/test_metalearning/pyMetaLearn/test_metalearner.py:46]: We probably shouldn't have all these asserts in the setup, but rather in their own unit test.

File: scripts/02_retrieve_metadata.py [hunk 18]
```diff
@@ -220,7 +225,10 @@ def main():
             )
 
             configuration_space = pipeline.get_configuration_space(
-                {"is_sparse": sparse, "task": task}
+                DummyDatamanager(
+                    info={"is_sparse": sparse, "task": task},
+                    feat_type={"A": "numerical"}
```
Comment (mfeurer) [inline scripts/02_retrieve_metadata.py:230]: Should there also be a `"B": "categorical"` here? Otherwise, categorical hyperparameters will not be part of the meta-learning.

**Review by mfeurer — COMMENTED** [review]
Only one thing left right now, should be easy to fix.

**Review by mfeurer — APPROVED** [review]
